File: CommonJava/src/main/java/org/khwu/string/KMP.java
package org.khwu.string;

import java.util.Arrays;

public class KMP {

    private final int R = 256;
    private final String pat;
    private final int[][] dfa;

    public KMP(String pat) {
        int len = pat.length();
        this.pat = pat;
        this.dfa = new int[R][len];

        dfa[pat.charAt(0)][0] = 1;
        for (int col = 1, pt = 0; col < len; col++) {
            for (int row = 0; col > 0 && row < R; row++) {
                dfa[row][col] = dfa[row][pt];
            }
            dfa[pat.charAt(col)][col] = col+1;
            pt = dfa[pat.charAt(col)][pt];
        }
    }

    public int search(String txt) {
        int m = pat.length();
        int n = txt.length();
        for (int i = 0, j = 0; i < n; i++) {
            j = dfa[txt.charAt(i)][j];
            if (j == m) return i-m+1;
        }
        return -1;
    }

    public static void main(String[] args) {
        String pat = "ABABAC";
        String txt = "AABABAC";
        KMP kmp = new KMP(pat);
        System.out.println(kmp.search(txt));
    }
}

File: python/string/kmp.py


class KMP(object):
	R = 256
	def __init__(self, pat):
		self.n = len(pat)
		self.dfa = [[0] * self.n for _ in range(KMP.R)]
		self.dfa[ord(pat[0])][0] = 1
		pt = 0
		for col in range(1, self.n):
			for row in range(KMP.R):
				self.dfa[row][col] = self.dfa[row][pt]
			self.dfa[ord(pat[col])][col] = col + 1
			pt = self.dfa[ord(pat[col])][pt]

		# for i, arr in enumerate(self.dfa):
		# 	if i < ord("a") or i > ord("z"):
		# 		continue
		# 	print("\t".join(map(str, arr)))

	def search(self, txt):
		i, j, m = 0, 0, len(txt)
		while i < m and j < self.n:
			j = self.dfa[ord(txt[i])][j]
			i += 1
		if j == self.n:
			return i - self.n
		return -1


if __name__ == "__main__":
	kmp = KMP("ababc")
	print(kmp.search("ababababb"))
